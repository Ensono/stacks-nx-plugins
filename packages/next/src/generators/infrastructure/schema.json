{
    "$schema": "http://json-schema.org/schema",
    "cli": "nx",
    "$id": "Next",
    "title": "Ensono Stacks Next Infrastructure Generator",
    "description": "The @ensono-stacks/next:infrastructure will add deployment infrastructure to an existing next project. Including generating a docker file, Heml chart and Terraform config",
    "type": "object",
    "properties": {
        "project": {
            "type": "string",
            "description": "The name of the project.",
            "alias": "p",
            "$default": {
                "$source": "projectName"
            },
            "x-prompt": "What is the name of the project for this component?"
        },
        "openTelemetry": {
            "type": "boolean",
            "description": "Auto-instrumentation for opentelemetry",
            "alias": "o",
            "$default": "false",
            "x-prompt": "Would you like auto-instrumentation for OpenTelemetry?"
        }
    },
    "required": ["project"]
}
