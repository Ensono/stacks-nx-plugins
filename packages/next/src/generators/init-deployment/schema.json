{
    "$schema": "http://json-schema.org/schema",
    "cli": "nx",
    "$id": "InitDeployment",
    "title": "Ensono Stacks Next Init Deployment Generator",
    "description": "The @ensono-stacks/next:init-deployment will add deployment infrastructure to an existing next project. Including generating a docker file, Heml chart and Terraform config",
    "type": "object",
    "properties": {
        "project": {
            "type": "string",
            "description": "The name of the project.",
            "alias": "p",
            "$default": {
                "$source": "projectName"
            },
            "x-prompt": "What is the name of the project for this component?"
        },
        "libraryName": {
            "type": "string",
            "description": "Name provided for generated helm chart library",
            "alias": "l",
            "default": "stacks-helm-chart",
            "x-prompt": "Would you like to specify a name for the generated helm chart library?"
        }
    },
    "required": ["project"]
}
