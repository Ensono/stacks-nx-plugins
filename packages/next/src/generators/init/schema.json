{
    "$schema": "http://json-schema.org/schema",
    "cli": "nx",
    "$id": "Next",
    "title": "",
    "type": "object",
    "properties": {
        "project": {
            "type": "string",
            "description": "The name of the project.",
            "alias": "p",
            "$default": {
                "$source": "projectName"
            },
            "x-prompt": "What is the name of the project for this component?"
        }
    },
    "required": ["project"]
}
