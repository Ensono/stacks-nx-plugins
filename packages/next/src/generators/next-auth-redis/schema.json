{
    "$schema": "http://json-schema.org/schema",
    "cli": "nx",
    "$id": "NextAuth",
    "title": "Ensono Stacks Next Auth Redis Generator",
    "description": "The @ensono-stacks/next:next-auth-redis generator will add Redis for session management in a Next-Auth app.",
    "type": "object",
    "properties": {
        "project": {
            "type": "string",
            "description": "The name of the project.",
            "alias": "p",
            "$default": {
                "$source": "projectName"
            },
            "x-prompt": "What is the name of the project?",
            "x-priority": "important"
        },
        "envVar": {
            "type": "string",
            "default": "REDIS_URL",
            "description": "Name of the env var that stores connection string for Redis",
            "x-prompt": "What is the name of the env var that stores connection string for Redis?",
            "pattern": "^[a-zA-Z_]+$"
        },
        "adapterName": {
            "type": "string",
            "default": "next-auth-redis",
            "description": "Name of the generated Redis adapter library"
        }
    },
    "required": ["project"]
}
