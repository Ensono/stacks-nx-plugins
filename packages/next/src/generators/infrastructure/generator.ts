import { StacksConfigError } from '@ensono-stacks/core';
import {
    joinPathFragments,
    readProjectConfiguration,
    GeneratorCallback,
    Tree,
} from '@nrwl/devkit';
import { runTasksInSerial } from '@nrwl/workspace/src/utilities/run-tasks-in-serial';
import chalk from 'chalk';

import { NextGeneratorSchema } from './schema';
import { addInfrastructure } from './utils/add-infrastructure';
import { formatFiles } from './utils/format-files';

export default async function infrastructureGenerator(
    tree: Tree,
    options: NextGeneratorSchema,
) {
    const tasks: GeneratorCallback[] = [];
    const project = readProjectConfiguration(tree, options.project);

    try {
        tasks.push(...addInfrastructure(tree, project));
    } catch (error) {
        if (error instanceof StacksConfigError) {
            console.warn(chalk.yellow`Missing Stacks configuration in nx.json`);
            console.log(
                '  Infrastructure code will not be applied now, but you can rerun this task when you update the configuration.',
            );
            console.log(
                '    nx g @ensono-stacks/next:infrastructure --project [project_name]',
            );
            return () => {};
        }

        throw error;
    }

    // exclude helm yaml files from initial format when generating the files
    await formatFiles(tree, [
        joinPathFragments(project.root, 'build', 'helm', '**', '*.yaml'),
    ]);

    return runTasksInSerial(...tasks);
}
