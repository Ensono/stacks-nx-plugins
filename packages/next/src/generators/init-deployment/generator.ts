import {
    formatFiles,
    executedDependantGenerator,
    hasGeneratorExecutedForProject,
    StacksError,
} from '@ensono-stacks/core';
import {
    joinPathFragments,
    readProjectConfiguration,
    GeneratorCallback,
    Tree,
} from '@nrwl/devkit';
import { runTasksInSerial } from '@nrwl/workspace/src/utilities/run-tasks-in-serial';
import chalk from 'chalk';

import { NextGeneratorSchema } from './schema';
import { addInfrastructure } from './utils/add-infrastructure';

export default async function initDeploymentGenerator(
    tree: Tree,
    options: NextGeneratorSchema,
) {
    if (
        !executedDependantGenerator(
            tree,
            'WorkspaceDeployment',
            options.project,
        )
    )
        return false;
    if (
        hasGeneratorExecutedForProject(
            tree,
            options.project,
            'NextInitDeployment',
            true,
        )
    )
        return false;

    const tasks: GeneratorCallback[] = [];
    const project = readProjectConfiguration(tree, options.project);

    try {
        tasks.push(...addInfrastructure(tree, project, options));
    } catch (error) {
        if (error instanceof StacksError) {
            console.warn(chalk.yellow`Missing Stacks configuration in nx.json`);
            console.log(
                '  Infrastructure code will not be applied now, but you can rerun this task when you update the configuration.',
            );
            console.log(
                '    nx g @ensono-stacks/next:init-deployment --project [project_name]',
            );
            return () => {};
        }

        throw error;
    }

    // exclude helm yaml files from initial format when generating the files
    await formatFiles(tree, [
        joinPathFragments(
            'libs',
            options.libraryName,
            'build',
            'helm',
            '**',
            '*.yaml',
        ),
    ]);

    return runTasksInSerial(...tasks);
}
