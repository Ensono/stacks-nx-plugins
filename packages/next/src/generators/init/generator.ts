import {
    formatFiles,
    formatFilesWithEslint,
    addCustomTestConfig,
    deploymentGeneratorMessage,
    hasGeneratorExecutedForProject,
    verifyPluginCanBeInstalled,
    tsMorphTree,
} from '@ensono-stacks/core';
import {
    GeneratorCallback,
    joinPathFragments,
    readProjectConfiguration,
    Tree,
    updateProjectConfiguration,
} from '@nx/devkit';
import { runTasksInSerial } from '@nx/workspace/src/utilities/run-tasks-in-serial';
import chalk from 'chalk';
import path from 'path';

import { NextGeneratorSchema } from './schema';
import { addEslint } from './utils/eslint';
import { eslintFix } from './utils/eslint-fix';
import {
    addReactAxeConfigToApp,
    addReactAxeDependency,
} from './utils/react-axe';
import { addStorybook } from './utils/storybook';
import updateTsConfig from './utils/tsconfig';

export default async function initGenerator(
    tree: Tree,
    options: NextGeneratorSchema,
) {
    verifyPluginCanBeInstalled(tree, options.project);

    if (hasGeneratorExecutedForProject(tree, options.project, 'NextInit'))
        return false;

    const tasks: GeneratorCallback[] = [];
    const project = readProjectConfiguration(tree, options.project);

    const update = { ...project };

    if (
        project.sourceRoot &&
        project.sourceRoot ===
            update.targets?.build?.configurations?.development?.outputPath
    ) {
        update.targets.build.configurations.development = {};
    }

    updateProjectConfiguration(tree, project.name, update);

    tasks.push(addEslint(tree, project.root), addStorybook(tree, options));

    const ciCoverageConfig = {
        ci: {
            collectCoverage: true,
            coverageReporters: ['text', 'html'],
            collectCoverageFrom: [
                './**/*.{js,jsx,ts,tsx}',
                './!**/.next/**',
                './!**/*.d.ts',
                './!**/*.config.*',
                './!**/_app.*',
            ],
            codeCoverage: true,
            ci: true,
        },
    };

    await addCustomTestConfig(
        tree,
        readProjectConfiguration(tree, options.project),
        project.name,
        ciCoverageConfig,
    );

    tasks.push(
        formatFilesWithEslint(options.project),
        addReactAxeDependency(tree),
    );

    // update tsconfig.json
    updateTsConfig(
        tree,
        project,
        path.join(project.sourceRoot, 'tsconfig.json'),
        ['next.config.js'],
    );

    // update tsconfig.spec.json
    updateTsConfig(
        tree,
        project,
        path.join(project.sourceRoot, 'tsconfig.spec.json'),
    );

    const morphTree = tsMorphTree(tree);

    console.info(chalk.green`Attempting to add configuration for react-axe`);
    addReactAxeConfigToApp(project, morphTree);
    console.info(
        chalk.green`continuing execution of next:init generator after attempting to add react-axe`,
    );
    eslintFix(project, tree);

    // exclude helm yaml files from initial format when generating the files
    await formatFiles(tree, [
        joinPathFragments(project.root, 'build', 'helm', '**', '*.yaml'),
    ]);

    return runTasksInSerial(...tasks, () =>
        deploymentGeneratorMessage(
            tree,
            `nx g @ensono-stacks/next:init-deployment --project ${options.project}`,
        ),
    );
}
