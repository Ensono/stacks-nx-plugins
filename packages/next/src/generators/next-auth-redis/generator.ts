/*
    TODO: https://dev.azure.com/amido-dev/Amido-Stacks/_workitems/edit/6236
    This generator has been removed from the generators.json because it needs work
*/
import {
    formatFilesWithEslint,
    createOrUpdateLocalEnv,
    hasGeneratorExecutedForProject,
    deploymentGeneratorMessage,
    verifyPluginCanBeInstalled,
    getNpmScope,
} from '@ensono-stacks/core';
import {
    joinPathFragments,
    readProjectConfiguration,
    Tree,
    logger,
    formatFiles,
    readNxJson,
    names,
    getWorkspaceLayout,
    generateFiles,
    runTasksInSerial,
} from '@nx/devkit';
import { libraryGenerator } from '@nx/js';
import path from 'path';

import { NextAuthRedisGeneratorSchema } from './schema';
import { DEFAULT_REDIS_URL } from './utils/constants';
import { installDependencies } from './utils/dependencies';
import { configureAdapter } from './utils/redis-adapter';

export default async function nextAuthRedisGenerator(
    tree: Tree,
    options: NextAuthRedisGeneratorSchema,
) {
    verifyPluginCanBeInstalled(tree, options.project);

    if (hasGeneratorExecutedForProject(tree, options.project, 'NextAuthRedis'))
        return false;

    const project = readProjectConfiguration(tree, options.project);

    const nextAuthApiFilePath = joinPathFragments(
        project.root,
        'src',
        'auth.ts',
    );
    if (!tree.exists(nextAuthApiFilePath)) {
        throw new Error(
            `Could not find ${nextAuthApiFilePath} . Did you configure Next-Auth for this project?`,
        );
    }

    const name = options.adapterName;

    const libraryName = names(name).fileName;
    const projectDirectory = libraryName;
    const projectRoot = `apps/${
        getWorkspaceLayout(tree).libsDir
    }/${projectDirectory}`;

    // generate the lib package
    await libraryGenerator(tree, {
        name: libraryName,
        directory: `apps/${libraryName}`,
        projectNameAndRootFormat: 'as-provided',
    });
    // delete the default generated lib folder
    const libraryDirectory = path.join(projectRoot, 'src');
    tree.delete(path.join(libraryDirectory, 'lib'));

    // add library files
    generateFiles(tree, path.join(__dirname, 'files'), projectRoot, {
        envVar: options.envVar,
        template: '',
    });

    configureAdapter(project, tree, {
        npmScope: `@${getNpmScope(tree)}`,
        libraryName,
        envVar: options.envVar,
    });

    createOrUpdateLocalEnv(project, tree, {
        [options.envVar]: DEFAULT_REDIS_URL,
    });

    await formatFiles(tree);

    return runTasksInSerial(
        installDependencies(tree),
        formatFilesWithEslint(options.project),
        () => {
            logger.warn(
                `Do not forget to update your .env.local environment variables with values.`,
            );
        },
        () =>
            deploymentGeneratorMessage(
                tree,
                `nx g @ensono-stacks/next:next-auth-redis-deployment --project ${options.project}`,
            ),
    );
}
