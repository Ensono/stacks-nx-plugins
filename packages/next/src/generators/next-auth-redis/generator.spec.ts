import { addStacksAttributes } from '@ensono-stacks/test';
import { Tree, readJson, updateJson } from '@nx/devkit';
import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing';
import { applicationGenerator } from '@nx/next';

import generator from './generator';
import { NextAuthRedisGeneratorSchema } from './schema';
import nextInitGenerator from '../init/generator';
import nextAuthGenerator from '../next-auth/generator';

jest.mock('@ensono-stacks/core', () => ({
    ...jest.requireActual('@ensono-stacks/core'),
    execAsync: jest.fn(),
    getCommandVersion: jest.fn(() => '1.0.0'),
}));

describe('next-auth-redis generator', () => {
    let appTree: Tree;
    const options: NextAuthRedisGeneratorSchema = {
        project: 'next-app',
        envVar: 'REDIS_URL',
        adapterName: 'next-auth-redis',
    };

    beforeEach(async () => {
        appTree = createTreeWithEmptyWorkspace();
        await applicationGenerator(appTree, {
            name: 'next-app',
            style: 'css',
        });
        addStacksAttributes(appTree, options.project);
        await nextInitGenerator(appTree, { project: 'next-app' });
        await nextAuthGenerator(appTree, {
            project: 'next-app',
            provider: 'none',
        });
    });

    it('should generate redis adapter lib', async () => {
        await generator(appTree, options);

        const nextAuthTs = appTree.read('next-app/src/auth.ts');

        console.log({ appTree });
        expect(nextAuthTs.toString()).toMatchSnapshot();

        expect(
            appTree.exists('apps/next-auth-redis/src/index.ts'),
        ).toBeTruthy();

        const localEnv = appTree.read('next-app/.env.local');
        expect(localEnv.toString()).toContain('REDIS_URL=localhost:6379');

        const packageJson = readJson(appTree, 'package.json');
        expect(Object.keys(packageJson.dependencies)).toEqual(
            expect.arrayContaining(['ioredis', 'uuid']),
        );
        expect(Object.keys(packageJson.devDependencies)).toEqual(
            expect.arrayContaining(['@types/uuid']),
        );
    });

    it('should generate redis adapter lib with custom env var name', async () => {
        await generator(appTree, {
            ...options,
            envVar: 'REDIS_CONNECTION_STRING',
        });

        const nextAuthTs = appTree.read('next-app/src/auth.ts');
        expect(nextAuthTs.toString()).toMatchSnapshot();

        const localEnv = appTree.read('next-app/.env.local');
        expect(localEnv.toString()).toContain(
            'REDIS_CONNECTION_STRING=localhost:6379',
        );
    });

    it('should generate redis adapter lib with custom name', async () => {
        await generator(appTree, {
            ...options,
            adapterName: 'redis-adapter-for-next-auth',
        });

        const nextAuthTs = appTree.read('next-app/src/auth.ts');
        expect(nextAuthTs.toString()).toMatchSnapshot();

        expect(
            appTree.exists('apps/redis-adapter-for-next-auth/src/index.ts'),
        ).toBeTruthy();
    });

    describe('executedGenerators', () => {
        beforeEach(async () => {
            await generator(appTree, {
                ...options,
                adapterName: 'redis-adapter-for-next-auth',
            });
        });

        it('should update nx.json and tag executed generator true', async () => {
            const nxJson = readJson(appTree, 'nx.json');

            expect(
                nxJson.stacks.executedGenerators.project[
                    options.project
                ].includes('NextAuthRedis'),
            ).toBe(true);
        });

        it('should return false from method and eit generator if already executed', async () => {
            const gen = await generator(appTree, {
                ...options,
            });

            expect(gen).toBe(false);
        });
    });
});
