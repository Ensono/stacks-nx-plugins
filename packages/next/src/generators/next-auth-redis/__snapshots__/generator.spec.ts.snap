// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`next-auth-redis generator should generate redis adapter lib 1`] = `
"import NextAuth from 'next-auth';
import { IORedisAdapter } from '@proj/next-auth-redis';
import { Redis } from 'ioredis';

export const { handlers, signIn, signOut, auth } = NextAuth({
  providers: [],
  adapter: IORedisAdapter(new Redis(process.env.REDIS_URL)),
});
"
`;

exports[`next-auth-redis generator should generate redis adapter lib with custom env var name 1`] = `
"import NextAuth from 'next-auth';
import { IORedisAdapter } from '@proj/next-auth-redis';
import { Redis } from 'ioredis';

export const { handlers, signIn, signOut, auth } = NextAuth({
  providers: [],
  adapter: IORedisAdapter(new Redis(process.env.REDIS_CONNECTION_STRING)),
});
"
`;

exports[`next-auth-redis generator should generate redis adapter lib with custom name 1`] = `
"import NextAuth from 'next-auth';
import { IORedisAdapter } from '@proj/redis-adapter-for-next-auth';
import { Redis } from 'ioredis';

export const { handlers, signIn, signOut, auth } = NextAuth({
  providers: [],
  adapter: IORedisAdapter(new Redis(process.env.REDIS_URL)),
});
"
`;
